src/etf.rs

```diff
@@ -228,3 +239,74 @@ impl DebugRegister for EtfMode {
     const ADDRESS: u32 = 0x28;
     const NAME: &'static str = "ETF_MODE";
 }
+
+/// Trace ID (a.k.a. ATID or trace source ID)
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Id(u8);
+impl From<u8> for Id {
+    fn from(id: u8) -> Self {
+        Self(id)
+    }
+}
+impl From<Id> for u8 {
+    fn from(id: Id) -> Self {
+        id.0
+    }
+}
+
+/// Formatted frame demultiplexer.
+/// Takes a reference to a 16 byte frame from the ETB/ETF or TPIU and
+/// reads source ID and bytes from it.
+#[derive(Copy, Clone, Debug)]
+pub struct Frame<'a> {
+    data: &'a [u8; 16],
+    idx: usize,
+    id: Id,
+}
+
+impl<'a> Frame<'a> {
+    pub fn new(data: &'a [u8; 16], id: Id) -> Self {
+        Self { data, id, idx: 0 }
+    }
+
+    pub fn id(&self) -> Id {
+        self.id
+    }
+
+    pub fn rewind(&mut self) {
+        self.idx = 0;
+    }
+}
+
+impl<'a> Iterator for &mut Frame<'a> {
+    type Item = (Id, u8);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // DDI0314H_coresight_components_trm (ARM DDI 0314H) 9.6.1,
+        // US20050039078A1,
+        // and others
+        if self.idx >= 15 {
+            return None;
+        }
+        let byte = self.data[self.idx];
+        let lsb = (self.data[15] >> (self.idx >> 1)) & 1;
+        let ret = match (self.idx & 1, byte & 1) {
+            (0, 0) => Some((self.id, byte | lsb)),
+            (0, 1) => {
+                let new_id = (byte >> 1).into();
+                let next_id = if lsb == 1 { self.id } else { new_id };
+                self.id = new_id;
+                if self.idx >= 14 {
+                    None
+                } else {
+                    self.idx += 1;
+                    Some((next_id, self.data[self.idx]))
+                }
+            }
+            (1, _) => Some((self.id, byte)),
+            _ => unreachable!(),
+        };
```

Comment:
The match on two bit flags seems a bit odd, since you end up with a weird unreachable match arm. I personally prefer the cascading if format to follow a logical flow of checking the two flags.

```suggestion
        let ret = if self.idx.bit(1) {
            // Odd indices of the frame always contain data associated with the previous ID.
            Some((self.id, byte))
        } else if byte.bit(1) {
            // Even bytes utilize the LSbit to indicate if they contain an ID or data. For an asserted LSbit, data is 
            // contained and the correct LSbit is stored at the end of the frame.
            Some((self.id, byte | lsb))
        } else {
            // Even bytes may also contain a new ID to swap to. In this case, the LSbit contained at the end of the
            // frame is used to indicate if the following data uses the new or old ID.
            let new_id = (byte >> 1).into();
            let next_id = if lsb == 1 { self.id } else { new_id };
            self.id = new_id;
            if self.idx >= 14 {
                None
            } else {
                self.idx += 1;
                Some((next_id, self.data[self.idx]))
            }
        };
```